src/support/small_set.h

```diff
@@ -34,10 +34,78 @@
 
 namespace wasm {
 
-template<typename T, size_t N, typename FlexibleSet> class SmallSetBase {
+template<typename T, size_t N> struct FixedStorageBase {
+  size_t used = 0;
+  std::array<T, N> storage;
+};
+
+template<typename T, size_t N>
+struct UnorderedFixedStorage : public FixedStorageBase<T, N> {
+  void insert(const T& x) {
+    assert(this->used < N);
+    this->storage[this->used++] = x;
+  }
+
+  void erase(const T& x) {
+    for (size_t i = 0; i < this->used; i++) {
+      if (this->storage[i] == x) {
+        // We found the item; erase it by moving the final item to replace it
+        // and truncating the size.
+        this->used--;
+        this->storage[i] = this->storage[this->used];
+        return;
+      }
+    }
+  }
+};
+
+template<typename T, size_t N>
+struct OrderedFixedStorage : public FixedStorageBase<T, N> {
+  void insert(const T& x) {
```

Comment:
We can save some work if we remove the precondition that the element is not already in the set, i.e. remove the call to `count` in `SmallSet::insert` and handle that in this method instead.

Reply:
Hmm, that might be worth doing, but how about separately? It could have been an optimization for the old code too so it seems separable.

And I'm not sure about the complexity - needs more thought. First, it would mean adding more code to `UnorderedFixedStorage::insert` (which will need a loop). Second, and more importantly, it makes `SmallSet::insert` more complicated if `*Storage::insert` can be a no-op, because then we need to use that information to determine whether to switch from fixed to flexible storage. In particular, we can't call `*Storage::insert` if we are using the maximum fixed size already, not unless we change the API to return an error in the case where the item does not already exist (in that case we'd go to flexible storage, but we can't first insert before doing that - we'd need to report an error instead).